pennylane/templates/base/single.py

```diff
@@ -0,0 +1,240 @@
+
+        If :math:`D` becomes large, the signature can be simplified by wrapping each entry in ``parameters``:
+
+        .. code-block:: python
+
+            @template
+            def mytemplate(pars, wires):
+                qml.Hadamard(wires=wires)
+                qml.RY(pars[0], wires=wires)
+                qml.RX(pars[1], wires=wires)
+
+            @qml.qnode(dev)
+            def circuit(pars):
+                Single(unitary=mytemplate, wires=[0,1,2], parameters=pars)
+                return qml.expval(qml.PauliZ(0))
+
+            print(circuit([[[1, 1]], [[2, 1]], [[0.1, 1]]]))
+
+        If the number of parameters for each wire does not match the template or gate, an error gets thrown:
+
+        .. code-block:: python
+
+                @template
+                def mytemplate(pars1, pars2, wires):
+                    qml.Hadamard(wires=wires)
+                    qml.RY(pars1, wires=wires)
+                    qml.RX(pars2, wires=wires)
+
+                @qml.qnode(dev)
+                def circuit(pars):
+                    Single(unitary=mytemplate, wires=[0, 1, 2], parameters=pars)
+                    return qml.expval(qml.PauliZ(0))
+
+
+        >>> circuit([1, 2, 3]))
+        TypeError: mytemplate() missing 1 required positional argument: 'pars2'
+
+        **Keyword arguments**
+
+        The unitary can be a template that takes additional keyword arguments.
+
+        .. code-block:: python
+
+            @template
+            def mytemplate(wires, h=True):
+                if h:
+                    qml.Hadamard(wires=wires)
+                qml.T(wires=wires)
+
+            @qml.qnode(dev)
+            def circuit(hadamard=None):
+                Single(unitary=mytemplate, wires=[0, 1, 2], kwargs={'h': hadamard})
+                return qml.expval(qml.PauliZ(0))
+
+            circuit(hadamard=False)
+
+    """
+
+    #########
+    # Input checks
+
+    wires = _check_wires(wires)
+
+    _check_type(parameters, [Iterable, type(None)], msg="'parameters' must be either of type None or "
+                                                        "Iterable; got {}".format(type(parameters)))
+#    _check_type(unitary, [list, FunctionType], msg="unitary must be a ``~.pennylane.operation.Operation`` "
+#                                                   "or a template function; got {}".format(type(unitary)))
+
+    if parameters is not None:
+        shape = _get_shape(parameters)
+        if shape[0] != len(wires):
+            raise ValueError("'parameters' must contain one entry for each of the {} wires; got shape {}"
+                             .format(len(wires), shape))
+        # repackage for consistent unpacking
+        if len(shape) == 1:
+            parameters = [[p] for p in parameters]
+    else:
+        parameters = [[] for _ in range(len(wires))]
+
+    #########
+
+    for w, p in zip(wires, parameters):
+        unitary(*p, wires=w, **kwargs)
```

Comment:
Wow, based on the docstring I was expected a very large function, but it's very concise! I suppose this is more about packaging up common use-cases that we have become familiar with, but users might not be.